data-prepper-core/src/main/java/com/amazon/dataprepper/plugin/PluginCreator.java

```diff
@@ -33,16 +40,56 @@
         }
     }
 
-    private <T> Constructor<?> getConstructor(final Class<T> pluginClass, final Class<?> pluginConfigurationType, final String pluginName) {
-        try {
-            return pluginClass.getConstructor(pluginConfigurationType);
-        } catch (final NoSuchMethodException ex) {
-            LOG.error("Data Prepper plugin requires a constructor with {} parameter;" +
-                            " Plugin {} with name {} is missing such constructor.", pluginConfigurationType,
-                    pluginClass.getSimpleName(), pluginName, ex);
-            throw new InvalidPluginDefinitionException(format("Data Prepper plugin requires a constructor with %s parameter;" +
-                            " Plugin %s with name %s is missing such constructor.", pluginConfigurationType,
-                    pluginClass.getSimpleName(), pluginName), ex);
+    private <T> Constructor<?> getConstructor(final Class<T> pluginClass, final Class<?> singleParameterType, final String pluginName) {
+
+        final Constructor<?>[] constructors = pluginClass.getConstructors();
+
+        final Optional<Constructor<?>> annotatedConstructor = getAnnotatedConstructor(pluginClass, constructors);
+        if(annotatedConstructor.isPresent())
+            return annotatedConstructor.get();
+
+
+        final Optional<Constructor<?>> singleParameterConstructor = Arrays.stream(constructors)
+                .filter(c -> Arrays.equals(c.getParameterTypes(), new Class[]{singleParameterType}))
+                .findFirst();
+
+        if(singleParameterConstructor.isPresent())
+            return singleParameterConstructor.get();
+
+        final Optional<Constructor<?>> defaultConstructor = Arrays.stream(constructors)
+                .filter(c -> c.getParameterTypes().length == 0)
+                .findFirst();
+
+
+        if(defaultConstructor.isPresent())
+            return defaultConstructor.get();
+
+        final String error =
+                String.format("Data Prepper plugin requires a constructor which is either: " +
+                        "1. Annotated with @DataPrepperPlugin; " +
+                        "2. Contains a single argument of type %s; or " +
```

Comment:
What happens if there are two separate single argument constructors? What is the failure mode?

What is your vision for v2.0? Do you plan on supporting all of these scenarios in the future as well? Would it be possible to move to an explicit model which requires an annotated constructor?

Reply:
> What happens if there are two separate single argument constructors?
Because the `@DataPrepperPlugin` allows configuring the `pluginConfigurationType`, the single argument constructor will only match that type. So it should only ever be what is defined by `pluginConfigurationType`.

> What is your vision for v2.0? Do you plan on supporting all of these scenarios in the future as well? Would it be possible to move to an explicit model which requires an annotated constructor?
For 2.0, I'd like to remove the single argument constructor. Thus, the constructor must be either annotated with `@DataPrepperConstructor` or be the default constructor.
We need to keep at least support for a single argument `PluginSetting` in 1.x since that is the current behavior. I could change this so that the only implicit single argument constructor is `PluginSetting` and disallow the POJO-based single-argument constructor. Would that help?

Reply:
@cmanning09 ,

I'm going to change this behavior so that the only single parameter constructor supported is `PluginSetting`. Thus, plugins which use a POJO for the configuration must provide the `@DataPrepperConstructor` annotation. I believe this should be simpler and is one fewer behavior to deprecate.

Reply:
I pushed a commit which, among other changes, now only permits a single-parameter constructor for `PluginSetting`. To use the custom POJO configuration, plugin authors must add the `@DataPrepperPluginConstructor` annotation.